java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -127,147 +133,180 @@ public void execute(EventMessage msg) {
             () -> {
-                LOG.info("Cannot find machine id for minion: " + minionId);
-                return false;
+                Optional<MinionServer> registeredMinionOpt = MinionServerFactory.findByMachineId(machineId);
+                Opt.consume(registeredMinionOpt,
+                    () -> {
+                        if (!duplicateMinionNamePresent(minionId)) {
+                            finalizeMinionRegistration(minionId, machineId, saltSSHProxyId, actKeyOverride, isSaltSSH);
+                        }
+                    },
+                    registeredMinion -> {
+                        updateAlreadyRegisteredInfo(minionId, registeredMinion);
+                        applySaltBootStates(minionId, registeredMinion, saltbootInitrd);
+                    });
             },
-            machineId -> {
-                boolean isReactivation = false;
-                ActivationKey ak = null;
-                ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
-                Optional<String> managmentKeyLabel = getManagementKeyLabelFromGrains(grains);
-                if (managmentKeyLabel.isPresent()) {
-                    ak = ActivationKeyFactory.lookupByKey(managmentKeyLabel.get());
-                    if (ak != null && ak.getKickstartSession() != null) {
-                        ak.getKickstartSession().markComplete("Installation completed.");
-                    }
-                    if (ak == null) {
-                        LOG.info("Outdated Management Key defined for " + minionId + ": " + managmentKeyLabel.get());
-                    }
-                    else if (ak.getServer() == null) {
-                        LOG.error("Management Key is not a reactivation key: " + managmentKeyLabel.get());
-                    }
-                    else {
-                        isReactivation = true;
-                    }
-                }
-
-                Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
-                if (!isReactivation &&
-                        checkIfMinionAlreadyRegistered(minionId, machineId, creator, isSaltSSH, grains)) {
-                    return true;
-                }
-                // Check if this minion id already exists
-                if (!isReactivation && duplicateMinionNamePresent(minionId)) {
-                    return false;
+            //Case-2 : Reactivation
+            mk -> {
+                if (isValidReactivationKey(mk, minionId)) {
+                    reactivateSystem(minionId, machineId, mk);
+                    finalizeMinionRegistration(minionId, machineId, saltSSHProxyId, actKeyOverride, isSaltSSH);
                 }
-                if (isReactivation) {
-                    reactivateSystem(minionId, machineId, creator, grains, Optional.ofNullable(ak));
+                else {
+                    MinionServerFactory.findByMachineId(machineId).ifPresent(rm -> {
```

Comment:
Sorry for being picky but I am still not convinced. Lines 199-201 are basically the same as 180+187-189: would it be possible to move `isValidReactivationKey` after line 155 so that we only enter the "Case-2 branch" if we have a key and we are also sure it's a reactivation key?

(please double check my logic, I might be missing something as well!!!)

Reply:
ha, no need to be sorry. This was really good. I didn't like my change for this corner case either. Your suggestion is good and it looks much cleaner now. I have pushed a commit.